### ISSUE TITLE
System.ArgumentOutOfRangeException

### ISSUE THREAD

**Issue by rameel** [agileobjects/ReadableExpressions#48]
```
An unhandled exception of type 'System.ArgumentOutOfRangeException' was thrown by the custom visualizer component in the process being debugged.
Length cannot be less than zero. (Parameter 'length')
   at System.String.Substring(Int32 startIndex, Int32 length)
   at AgileObjects.ReadableExpressions.Extensions.PublicTypeExtensions.GetGenericTypeVariableName(String variableName, Type namingType, TranslationSettings settings)
   at AgileObjects.ReadableExpressions.Extensions.PublicTypeExtensions.GetVariableName(Type type, TranslationSettings settings)
   at AgileObjects.ReadableExpressions.Extensions.PublicTypeExtensions.GetVariableNameInCamelCase(Type type, TranslationSettings settings)
   at AgileObjects.ReadableExpressions.Translations.ParameterTranslation.UnnamedParameterTranslation..ctor(ParameterExpression parameter, ITranslationContext context)
   at AgileObjects.ReadableExpressions.Translations.ParameterTranslation.For(ParameterExpression parameter, ITranslationContext context)
   at AgileObjects.ReadableExpressions.Translations.TranslationTree.GetTranslationFor(Expression expression)
   at AgileObjects.ReadableExpressions.Translations.TranslationTree.AgileObjects.ReadableExpressions.Translations.Interfaces.ITranslationContext.GetCodeBlockTranslationFor(Expression expression)
   at AgileObjects.ReadableExpressions.Translations.AssignmentTranslation..ctor(BinaryExpression assignment, ITranslationContext context)
   at AgileObjects.ReadableExpressions.Translations.TranslationTree.GetTranslationFor(Expression expression)
   at AgileObjects.ReadableExpressions.Translations.TranslationTree.AgileObjects.ReadableExpressions.Translations.Interfaces.ITranslationContext.GetTranslationFor(Expression expression)
   at AgileObjects.ReadableExpressions.Translations.BlockTranslation.BlockStatementTranslation..ctor(Expression expression, ITranslationContext context)
   at AgileObjects.ReadableExpressions.Translations.BlockTranslation.BlockAssignmentStatementTranslation..ctor(BinaryExpression assignment, ITranslationContext context)
   at AgileObjects.ReadableExpressions.Translations.BlockTranslation.GetBlockStatements(BlockExpression block, ITranslationContext context, Boolean& hasMultiStatementStatement, Int32& estimatedStatementsSize, Boolean& hasGoto)
   at AgileObjects.ReadableExpressions.Translations.BlockTranslation..ctor(BlockExpression block, ITranslationContext context)
   at AgileObjects.ReadableExpressions.Translations.TranslationTree.GetTranslationFor(Expression expression)
   at AgileObjects.ReadableExpressions.Translations.TranslationTree.AgileObjects.ReadableExpressions.Translations.Interfaces.ITranslationContext.GetCodeBlockTranslationFor(Expression expression)
   at AgileObjects.ReadableExpressions.Translations.ConditionalTranslation.TernaryTranslation..ctor(ConditionalExpression conditional, ITranslationContext context)
   at AgileObjects.ReadableExpressions.Translations.ConditionalTranslation.For(ConditionalExpression conditional, ITranslationContext context)
   at AgileObjects.ReadableExpressions.Translations.TranslationTree.GetTranslationFor(Expression expression)
   at AgileObjects.ReadableExpressions.Translations.TranslationTree.AgileObjects.ReadableExpressions.Translations.Interfaces.ITranslationContext.GetCodeBlockTranslationFor(Expression expression)
   at AgileObjects.ReadableExpressions.Translations.ConditionalTranslation.TernaryTranslation..ctor(ConditionalExpression conditional, ITranslationContext context)
   at AgileObjects.ReadableExpressions.Translations.ConditionalTranslation.For(ConditionalExpression conditional, ITranslationContext context)
   at AgileObjects.ReadableExpressions.Translations.TranslationTree.GetTranslationFor(Expression expression)
   at AgileObjects.ReadableExpressions.Translations.TranslationTree.AgileObjects.ReadableExpressions.Translations.Interfaces.ITranslationContext.GetTranslationFor(Expression expression)
   at AgileObjects.ReadableExpressions.Translations.BlockTranslation.BlockStatementTranslation..ctor(Expression expression, ITranslationContext context)
   at AgileObjects.ReadableExpressions.Translations.BlockTranslation.GetBlockStatements(BlockExpression block, ITranslationContext context, Boolean& hasMultiStatementStatement, Int32& estimatedStatementsSize, Boolean& hasGoto)
   at AgileObjects.ReadableExpressions.Translations.BlockTranslation..ctor(BlockExpression block, ITranslationContext context)
   at AgileObjects.ReadableExpressions.Translations.TranslationTree.GetTranslationFor(Expression expression)
   at AgileObjects.ReadableExpressions.Translations.TranslationTree.AgileObjects.ReadableExpressions.Translations.Interfaces.ITranslationContext.GetCodeBlockTranslationFor(Expression expression)
   at AgileObjects.ReadableExpressions.Translations.LambdaTranslation..ctor(LambdaExpression lambda, ITranslationContext context)
   at AgileObjects.ReadableExpressions.Translations.TranslationTree.GetTranslationFor(Expression expression)
   at AgileObjects.ReadableExpressions.Translations.TranslationTree..ctor(Expression expression, TranslationSettings settings)
   at AgileObjects.ReadableExpressions.ExpressionExtensions.ToReadableString(Expression expression, Func`2 configuration)
   at AgileObjects.ReadableExpressions.Visualizers.Core.ExpressionVisualizerObjectSource.GetData(Object target, Stream outgoingData, Action`2 serializer)
   at AgileObjects.ReadableExpressions.Visualizers.Vs16ExpressionVisualizerObjectSource.GetData(Object target, Stream outgoingData)
   at Microsoft.VisualStudio.DebuggerVisualizers.DebuggeeSide.Impl.ClrCustomVisualizerDebuggeeHost.GetData(Object visualizedObject)
```

**Comment by sdanyliv** [comment]
@rameel, for fast solving problem, i would suggest to check exception by yourself. It is also a separate library, so you can quickly find problem when test with expression which fail in your production code.

**Comment by rameel** [comment]
The exception is thrown when trying to generate a variable name for an inner class of a generic type, for example: ``ImmutableDictionary<int, int>.Builder``.

The following method returns a simple name ``Builder`` for this type (see ``namingType.Name``)
```cs
private static string GetBaseVariableName(Type namingType, TranslationSettings translationSettings)
    => namingType.IsPrimitive()
          ? namingType.GetFriendlyName(translationSettings)
          : namingType.Name;
```
and afterwards in the method ``GetGenericTypeVariableName`` we are trying to get find a symbol that does not exist
```cs
variableName = variableName.Substring(0, variableName.IndexOf('`'));
```

**Comment by SteveWilkes** [comment]
Thanks for reporting this - it's fixed in the current v2.3.2 release branch. Release to follow!

Steve

**Comment by SteveWilkes** [comment]
This is fixed as of v2.4, which is now [on NuGet](https://www.nuget.org/packages/AgileObjects.ReadableExpressions). Thanks for the feedback!
